warehouse/utils/utils_test.go

```diff
@@ -1,139 +1,917 @@
+			region:     "us-west-1",
+		},
+
+		{
+			location:   "https://s3.amazonaws.com/test-bucket/test-object.csv",
+			s3Location: "s3://test-bucket/test-object.csv",
+			region:     "",
+		},
+		{
+			location:   "https://s3.any-region.amazonaws.com/test-bucket/test-object.csv",
+			s3Location: "s3://test-bucket/test-object.csv",
+			region:     "any-region",
+		},
+		{
+			location:   "https://s3.amazonaws.com/my.test-bucket/test-object.csv",
+			s3Location: "s3://my.test-bucket/test-object.csv",
+			region:     "",
+		},
+		{
+			location:   "https://s3.us-west-1.amazonaws.com/my.test-bucket/test-object.csv",
+			s3Location: "s3://my.test-bucket/test-object.csv",
+			region:     "us-west-1",
+		},
+		{
+			location:   "https://s3.amazonaws.com/bucket.with.a.dot/test-object.csv",
+			s3Location: "s3://bucket.with.a.dot/test-object.csv",
+			region:     "",
+		},
+		{
+			location:   "https://s3.amazonaws.com/s3.rudderstack/test-object.csv",
+			s3Location: "s3://s3.rudderstack/test-object.csv",
+			region:     "",
+		},
+	}
+
+	for idx, input := range inputs {
+		s3Location, region := GetS3Location(input.location)
+		if s3Location != input.s3Location {
+			t.Errorf("got %q want %q input %d", s3Location, input.s3Location, idx)
+		}
+		if region != input.region {
+			t.Errorf("got %q want %q input %d", region, input.region, idx)
+		}
+	}
+}
+
+func TestCaptureRegexGroup(t *testing.T) {
+	inputs := []struct {
+		regex   string
+		pattern string
+		groups  map[string]string
+	}{
+		{
+			regex:   "https?://s3([.-](?P<region>[^.]+))?.amazonaws.com/(?P<bucket>[^/]+)/(?P<keyname>.*)",
+			pattern: "https://s3.amazonaws.com/bucket.with.a.dot/keyname",
+			groups: map[string]string{
+				"bucket":  "bucket.with.a.dot",
+				"keyname": "keyname",
+				"region":  "",
+			},
+		},
+		{
+			regex:   "https?://s3([.-](?P<region>[^.]+))?.amazonaws.com/(?P<bucket>[^/]+)/(?P<keyname>.*)",
+			pattern: "https://s3.us-east.amazonaws.com/bucket.with.a.dot/keyname",
+			groups: map[string]string{
+				"bucket":  "bucket.with.a.dot",
+				"keyname": "keyname",
+				"region":  "us-east",
+			},
+		},
+		{
+			regex:   "https?://(?P<bucket>[^/]+).s3([.-](?P<region>[^.]+))?.amazonaws.com/(?P<keyname>.*)",
+			pattern: "https://bucket.with.a.dot.s3.amazonaws.com/keyname",
+			groups: map[string]string{
+				"bucket":  "bucket.with.a.dot",
+				"keyname": "keyname",
+				"region":  "",
+			},
+		},
+		{
+			regex:   "https?://(?P<bucket>[^/]+).s3([.-](?P<region>[^.]+))?.amazonaws.com/(?P<keyname>.*)",
+			pattern: "https://bucket.with.a.dot.s3.amazonaws.com/keyname",
+			groups: map[string]string{
+				"bucket":  "bucket.with.a.dot",
+				"keyname": "keyname",
+				"region":  "",
+			},
+		},
+	}
+	for idx, input := range inputs {
+		got, err := CaptureRegexGroup(input.regex, input.pattern)
+		assertNoError(t, err)
+		if !reflect.DeepEqual(got, input.groups) {
+			t.Errorf("got %#v want %#v input %d", got, input.groups, idx)
+		}
```

Comment:
There is a very light library `testtify/require` which can be used to perform these assertions. Will streamline the statements I think. Wdyt ?